overlord/servicestate/quota_handlers_test.go

```diff
@@ -101,6 +101,137 @@ func (s *quotaHandlersSuite) TestDoQuotaControlCreate(c *C) {
+		// after task restart
+		systemctlCallsForServiceRestart("test-snap"),
+	))
+	defer r()
+
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	// setup the snap so it exists
+	snapstate.Set(s.state, "test-snap", s.testSnapState)
+	snaptest.MockSnapCurrent(c, testYaml, s.testSnapSideInfo)
+
+	// make a fake task
+	t := st.NewTask("create-quota", "...")
+
+	qcs := []servicestate.QuotaControlAction{
+		{
+			Action:      "create",
+			QuotaName:   "foo-group",
+			MemoryLimit: quantity.SizeGiB,
+			AddSnaps:    []string{"test-snap"},
+		},
+	}
+
+	t.Set("quota-control-actions", &qcs)
+
+	expectedQuotaState := map[string]quotaGroupState{
+		"foo-group": {
+			MemoryLimit: quantity.SizeGiB,
+			Snaps:       []string{"test-snap"},
+		},
+	}
+
+	st.Unlock()
+	err := s.o.ServiceManager().DoQuotaControl(t, nil)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(t.Status(), Equals, state.DoneStatus)
+
+	checkQuotaState(c, st, expectedQuotaState)
+
+	t.SetStatus(state.DoingStatus)
+
+	st.Unlock()
+	err = s.o.ServiceManager().DoQuotaControl(t, nil)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(t.Status(), Equals, state.DoneStatus)
+
+	checkQuotaState(c, st, expectedQuotaState)
+}
+
+func (s *quotaHandlersSuite) TestQuotaStateAlreadyUpdatedBehavior(c *C) {
+	// test a situation where because of restart the task is reentered
+	r := s.mockSystemctlCalls(c, join(
+		// doQuotaControl handler to create the group
+		systemctlCallsForCreateQuota("foo-group", "test-snap"),
+	))
+	defer r()
+
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	// setup the snap so it exists
+	snapstate.Set(s.state, "test-snap", s.testSnapState)
+	snaptest.MockSnapCurrent(c, testYaml, s.testSnapSideInfo)
+
+	// make a fake task
+	t := st.NewTask("create-quota", "...")
+
+	qcs := []servicestate.QuotaControlAction{
+		{
+			Action:      "create",
+			QuotaName:   "foo-group",
+			MemoryLimit: quantity.SizeGiB,
+			AddSnaps:    []string{"test-snap"},
+		},
+	}
+
+	t.Set("quota-control-actions", &qcs)
+
+	st.Unlock()
+	err := s.o.ServiceManager().DoQuotaControl(t, nil)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(t.Status(), Equals, state.DoneStatus)
+	t.SetStatus(state.DoingStatus)
+
+	updated, appsToRestart, err := servicestate.QuotaStateAlreadyUpdated(t)
+	c.Assert(err, IsNil)
+	c.Check(updated, Equals, true)
+	c.Assert(appsToRestart, HasLen, 1)
+	for info, apps := range appsToRestart {
+		c.Check(info.InstanceName(), Equals, "test-snap")
+		c.Assert(apps, HasLen, 1)
+		c.Check(apps[0], Equals, info.Apps["svc1"])
+	}
+
+	// rebooted
+	r = servicestate.MockOsutilBootID("other-boot")
+	defer r()
+
+	updated, appsToRestart, err = servicestate.QuotaStateAlreadyUpdated(t)
+	c.Assert(err, IsNil)
+	c.Check(updated, Equals, true)
+	c.Check(appsToRestart, HasLen, 0)
+	r()
+
+	// restored
+	_, appsToRestart, err = servicestate.QuotaStateAlreadyUpdated(t)
+	c.Assert(err, IsNil)
+	c.Check(appsToRestart, HasLen, 1)
+
+	// snap gone missing
```

Comment:
Nitpick, "snap gone" or "snap missing"?

Reply:
"snap went missing" would be my preference

Reply:
I changed this in https://github.com/snapcore/snapd/pull/10484